### DISCUSSION TITLE
Graphcache - Invalidate Document

### DISCUSSION THREAD

**Discussion by ragokan** [urql-graphql/urql#2273] [Q&A] [answered]
Hello, I read the docs but I don't see any way to invalidate document.
When I use *login* or *logout* mutation, I want to invalidate *getUser* query but their *__typename*s are different.

For example when I do this;

```ts
 updates: {
    Mutation: {
      login: ({ login }, _, cache) => {
        cache.invalidate({ __typename: "User", id: login.id });
      },
      logout: (_, __, cache) => cache.updateQuery({ query: GetUserDocument }, (data: Query["me"]) => null),
    },
  },
  ```
  Neither does work correctly.
  
  In App.vue;
  ```html
<script lang="ts" setup>
import { useGetUserQuery } from "./hooks/graphql";
const { data: user } = useGetUserQuery();

watch( () => user.value, () => console.log("user is changed"));
watch( () => user.value, () => console.log("user is changed: deep"), { deep: true } );
// They are called only at the beginning.
</script>
  ```
  
  In Login component
  ```ts
  const submit = () => {
   loginMutation({ data }); // does not trigger useGetUserQuery again.
};

const logout = () => {
   logoutMutation({});  // does not change **user** of **useGetUserQuery** 
};
  ```

**Comment by kitten** [comment, marked as answer]
I don quite know what your query is obviously, but when you login, you can't invalidate what's already being returned. So, if your login mutation returns a `User` there's in theory no need to invalidate anything for the query that contains the user, unless the ID changes, which means you could use `cache.link` to update the field

If you're only after updating a single field, e.g. `Query.user` without arguments, let's say then you'd likely want to use: `cache.invalidate('Query', 'user')`
If you're only looking to update said field you could use `cache.link` on login to make sure the key is up-to-date using: `cache.link('Query', 'user', login)`

**Reply by ragokan** [reply]
> I don quite know what your query is obviously, but when you login, you can't invalidate what's already being returned. So, if your login mutation returns a `User` there's in theory no need to invalidate anything for the query that contains the user, unless the ID changes, which means you could use `cache.link` to update the field
> 
> If you're only after updating a single field, e.g. `Query.user` without arguments, let's say then you'd likely want to use: `cache.invalidate('Query', 'user')` If you're only looking to update said field you could use `cache.link` on login to make sure the key is up-to-date using: `cache.link('Query', 'user', login)`

Heyy, doing like this works very well:
```ts
login: (_, __, cache) => cache.invalidate("Query", "me"),
logout: (_, __, cache) => cache.invalidate("Query", "me"),
```
Thank you so much.

Btw, what I was doing is this:
In main file, I watch the userQuery (me), whenever it changes I set the user.

Login mutation -> only gets success:true and session cookie
Logout mutation -> only gets success:true and destroys session cookie

So, whenever I get the cookie, I do another request and call get user query(me).
